File: Algorithms/Graph Theory/Prim's (MST) : Special Subtree.php
<?php

/*
 * Complete the 'prims' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts following parameters:
 *  1. INTEGER n
 *  2. 2D_INTEGER_ARRAY edges
 *  3. INTEGER start
 */

function prims($n, $edges, $start) {
    # Write your code here
    $adj = [];
    for ($i = 1; $i <= $n; $i++) {
        $adj[$i] = [];
    }

    foreach ($edges as $edge) {
        list($u, $v, $w) = $edge;
        $adj[$u][] = [$w, $v];
        $adj[$v][] = [$w, $u];
    }

    $pq = new SplPriorityQueue();

    $pq->setExtractFlags(SplPriorityQueue::EXTR_BOTH);

    $pq->insert($start, 0);

    $visited = array_fill(1, $n, False);
    $mst_weight = 0;

    while (!$pq->isEmpty()) {
        $curr = $pq->extract();
        $weight = -$curr['priority'];
        $node = $curr['data'];

        if ($visited[$node]) {
            continue;
        }

        $mst_weight += $weight;
        $visited[$node] = True;

        foreach ($adj[$node] as $neighbor) {
            list($next_weight, $next_node) = $neighbor;
            if (!$visited[$next_node]) {
                $pq->insert($next_node, -$next_weight);
            }
        }
    }

    return $mst_weight;
}

$fptr = fopen(getenv("OUTPUT_PATH"), "w");

$first_multiple_input = explode(' ', rtrim(fgets(STDIN)));

$n = intval($first_multiple_input[0]);
$m = intval($first_multiple_input[1]);

$edges = array();

for ($i = 0; $i < $m; $i++) {
    $edges_temp = rtrim(fgets(STDIN));

    $edges[] = array_map('intval', preg_split('/ /', $edges_temp, -1, PREG_SPLIT_NO_EMPTY));
}

$start = intval(trim(fgets(STDIN)));
$result = prims($n, $edges, $start);

fwrite($fptr, $result . "\n");
fclose($fptr);

File: Algorithms/Graph Theory/Prim's (MST) : Special Subtree.py
import os
import heapq

#
# Complete the 'prims' function below.
#
# The function is expected to return an INTEGER.
# The function accepts following parameters:
#  1. INTEGER n
#  2. 2D_INTEGER_ARRAY edges
#  3. INTEGER start
#

def prims(n, edges, start):
    # Write your code here
    adj = {i: [] for i in range(1, n + 1)}
    for u, v, w in edges:
        adj[u].append((w, v))
        adj[v].append((w, u))

    min_heap = [(0, start)]
    visited = set()
    mst_weight = 0

    while min_heap:
        weight, node = heapq.heappop(min_heap)

        if node in visited:
            continue
        mst_weight += weight
        visited.add(node)

        for next_weight, next_node in adj[node]:
            if next_node not in visited:
                heapq.heappush(min_heap, (next_weight, next_node))

    return mst_weight


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    first_multiple_input = input().rstrip().split()

    n = int(first_multiple_input[0])
    m = int(first_multiple_input[1])

    edges = []

    for _ in range(m):
        edges.append(list(map(int, input().rstrip().split())))

    start = int(input().strip())
    result = prims(n, edges, start)

    fptr.write(str(result) + '\n')
    fptr.close()
